### PR TITLE
Rename the .impl property and ensure all files use unwrap as needed

### PR DESCRIPTION
**Pull request by arv**
This is in preparation of allowing different strategies for wrapping

### CODE REVIEW

**Comment by arv** [comment]
@sorvell @sjmiles @azakus 

Heads up! I think you might have been using `.impl` before.

**Comment by arv** [comment]
Ping

**Comment by arv** [comment]
Actually, lets wait with this one a little. Let's land it (if you are OK with it) just after next release branch.

**Comment by dfreedm** [comment]
@arv, What's the motivation for changing the wrapping strategy?
Seems like a pretty big change.

**Comment by arv** [comment]
How is this a big change? It just renames `.impl`.

I want to potentially switch to not using wrappers and just replace the properties where it is supported.

Also, this change enables us to do cross frame DOM access if we want to support that.

**Comment by arv** [comment]
How about just renaming it to `impl_`?

**Comment by arv** [comment]
@sjmiles @sorvell 

Here is the change I was talking about.

As you can see `isWrapper` and `isNative` are now just checking for the presence of this property, allowing cross frame uwrapping.

Wrapping an object from a different frame would need some more work to find the correct wrapper. I have some ideas here too.

**Comment by arv** [comment]
@jmesserly What do you think about this? I still think this is a good change.

File: src/wrappers.js [hunk 1]
```diff
@@ -110,23 +110,31 @@ window.ShadowDOMPolyfill = {};
     return /^\w[a-zA-Z_0-9]*$/.test(name);
   }
 
+  // The name of the implementation property is intentionally hard to
+  // remember. Unfortunately, browsers are slower doing obj[expr] than
+  // obj.foo so we resort to repeat this ugly name. This ugly name is never
+  // used outside of this file though.q
```
Comment (jmesserly) [inline src/wrappers.js]: extra "q" at the end of this comment

File: src/wrappers.js [hunk 2]
```diff
@@ -110,23 +110,31 @@ window.ShadowDOMPolyfill = {};
...
+
   function getGetter(name) {
     return hasEval && isIdentifierName(name) ?
-        new Function('return this.impl.' + name) :
-        function() { return this.impl[name]; };
+        new Function('return this.__impl4cf1e782hg32d07__.' + name) :
```
Comment (jmesserly) [inline src/wrappers.js]: really crazy would be to regenerate this per build, but that seems like serious overkill :)

File: src/wrappers.js [hunk 3]
```diff
@@ -250,38 +258,12 @@ window.ShadowDOMPolyfill = {};
...
-
   function isWrapper(object) {
-    return object instanceof wrappers.EventTarget ||
-           object instanceof wrappers.Event ||
-           object instanceof wrappers.Range ||
-           object instanceof wrappers.DOMImplementation ||
-           object instanceof wrappers.CanvasRenderingContext2D ||
-           wrappers.WebGLRenderingContext &&
-               object instanceof wrappers.WebGLRenderingContext;
+    return object && object.__impl4cf1e782hg32d07__;
```
Comment (jmesserly) [inline src/wrappers.js]: very nice!

**Comment by jmesserly** [comment]
lgtm. Nice change.

**Comment by arv** [comment]
@azakus Are you still concerned about this change?

**Comment by dfreedm** [comment]
No, I think this is fine. Needs a rebaseline though.

**Comment by arv** [comment]
Rebased and updated new code.

PTAL

**Comment by jmesserly** [comment]
LGTM :)

**Comment by nazar-pc** [comment]
My code not working after update to 0.4.0 and I think this pull request caused that.
For example, I've used `.impl` in this case:

``` javascript
player_element = document.createElement('audio');
context = new AudioContext;
source = context.createMediaElementSource(player_element.impl);
```

Because if I don't I get:

> TypeError: Argument 1 of AudioContext.createMediaElementSource does not implement interface HTMLMediaElement.

I have no idea how to fix this now. Can I rely on that weird `.__impl4cf1e782hg__` in future versions?

**Comment by nazar-pc** [comment]
BTW, `does not implement interface HTMLMediaElement` - why not?

**Comment by jmesserly** [comment]
You can use ShadowDOMPolyfill.unwrap(player_element) ... Needless to say,
don't use **impl***   :)

(edit: formatting)

**Comment by nazar-pc** [comment]
I think this is too far from obvious and will cause huge compatibility issues when using with existing code.

**Comment by arv** [comment]
@nazar-pc This was actually done to reduce the compatibility issues. Using wrap/unwrap allows us to do the right thing on all browsers.

**Pull request merged**